File: apps/android/app/src/main/java/com/biupiu/rndos/RuntimeSession.kt
package com.biupiu.rndos

data class RuntimeSession(
    val client: String = "android",
    val subscriberId: String,
    val tier: String,
    val status: String,
    val entitlements: List<String>
)

data class RuntimeRouteState(
    val route: String,
    val enterable: Boolean,
    val reason: String
)

object BiupiuRuntime {
    fun resolveRoute(session: RuntimeSession, route: String): RuntimeRouteState {
        if (session.status != "active") {
            return RuntimeRouteState(route, false, "INACTIVE_ACCOUNT")
        }
        if (route == "MAIN_HUB" || route == "RND_OS") {
            return RuntimeRouteState(route, true, "AUTHORIZED")
        }
        val entitlement = when (route) {
            "SMART_FARMING" -> "SMART_FARMING"
            "SMART_METAL_WORKSHOP" -> "SMART_METAL_WORKSHOP"
            else -> return RuntimeRouteState(route, false, "NO_ENTITLEMENT")
        }
        return RuntimeRouteState(
            route,
            session.entitlements.contains(entitlement),
            if (session.entitlements.contains(entitlement)) "AUTHORIZED" else "NO_ENTITLEMENT"
        )
    }
}

File: apps/windows/RuntimeSession.cs
namespace Biupiu.Desktop;

public sealed record RuntimeSession(
    string SubscriberId,
    string Tier,
    string Status,
    IReadOnlyList<string> Entitlements);

public sealed record RuntimeRouteState(
    string Route,
    bool Enterable,
    string Reason);

public static class BiupiuRuntime
{
    public static RuntimeRouteState ResolveRoute(RuntimeSession session, string route)
    {
        if (session.Status != "active")
            return new(route, false, "INACTIVE_ACCOUNT");

        if (route is "MAIN_HUB" or "RND_OS")
            return new(route, true, "AUTHORIZED");

        var entitlement = route switch
        {
            "SMART_FARMING" => "SMART_FARMING",
            "SMART_METAL_WORKSHOP" => "SMART_METAL_WORKSHOP",
            _ => null
        };

        if (entitlement is null)
            return new(route, false, "NO_ENTITLEMENT");

        var allowed = session.Entitlements.Contains(entitlement);
        return new(route, allowed, allowed ? "AUTHORIZED" : "NO_ENTITLEMENT");
    }
}
